src/lib.rs

```diff
@@ -101,14 +142,30 @@ fn home_dir_inner() -> Option<PathBuf> {
 /// }
 /// ```
 pub fn cargo_home() -> io::Result<PathBuf> {
-    let cwd = env::current_dir()?;
-    cargo_home_with_cwd(&cwd)
+    cargo_home_from(&OS_ENV)
+}
+
+/// Variant of cargo_home where the environment source is parameterised. This is
+/// specifically to support in-process testing scenarios as environment
+/// variables and user home metadata are normally process global state. See the
+/// OsEnv trait.
+pub fn cargo_home_from(env: &dyn Env) -> io::Result<PathBuf> {
+    let cwd = env.current_dir()?;
+    cargo_home_with_cwd_from(env, &cwd)
 }
 
 /// Returns the storage directory used by Cargo within `cwd`.
 /// For more details, see [`cargo_home`](fn.cargo_home.html).
 pub fn cargo_home_with_cwd(cwd: &Path) -> io::Result<PathBuf> {
-    match env::var_os("CARGO_HOME").filter(|h| !h.is_empty()) {
+    cargo_home_with_cwd_from(&OS_ENV, cwd)
+}
+
+/// Variant of cargo_home_with_cwd where the environment source is
+/// parameterised. This is specifically to support in-process testing scenarios
```

Comment:
Nitpick: I prefer American English!
```suggestion
/// parameterized. This is specifically to support in-process testing scenarios
```

Reply:
twitch. ok.

Reply:
Thanks. Rustc prefer American English so I want to follow its choice.